### PR TITLE
add basic ip centric access control on routes

### PR DESCRIPTION
**Pull request by leprechau**
I've verified and tested this locally with good results.  Running it in our development environment for testing as well.

### CODE REVIEW

**Comment by leprechau** [comment]
Oops, forgot to link this to #439

File: route/access_rules.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+func (t *Target) AccessDeniedHTTP(r *http.Request) bool {
+	host, _, err := net.SplitHostPort(r.RemoteAddr)
+
+	if err != nil {
+		log.Printf("[ERROR] Failed to get host from RemoteAddr %s: %s",
+			r.RemoteAddr, err.Error())
+		return false
+	}
+
+	// prefer xff header if set
```
Comment (leprechau) [inline route/access_rules.go]: I have some mixed feelings about this section.  In the case where fabio may be behind another proxy you would want to honor `XFF` but then what about the actual originator?  Would it be better to check both the `XFF` (if present and different) as well as the actual source and return `true` from this function if either the `XFF` or `RemoteAddr` pass returns `true`?

File: route/access_rules.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	if xff := r.Header.Get("X-Forwarded-For"); xff != "" && xff != host {
+		host = xff
+	}
+
+	// currently only one function - more may be added in the future
+	return t.denyByIP(net.ParseIP(host))
+}
```
Comment (leprechau) [inline route/access_rules.go]: I have mixed feelings about this section.  In the event that fabio is deployed behind another load balancer you would want to check the `XFF` header but then what about the actual source?  Would it be better to check both the `XFF` (if present and different) as well as the actual `RemoteAddr` and return `true` (deny) if either of those checks return `true` on their own?

**Comment by leprechau** [comment]
It seems the tests are randomly failing with a timeout around some consul interaction.  I've seen the different go versions succeed and fail at different times with the same code by just amending the commit and doing force push.

File: route/access_rules.go (hunk 2 above)
Comment (leprechau) [reply route/access_rules.go]: Yes, I definitely don't like this.  It's an open exploit as is since there is no validation on who is allowed to pass an `XFF` header to fabio.

Example with `allow=ip:172.16.0.0/12` and a client that is not sourcing from that network.

```
$ curl -H 'X-Forwarded-For: 172.16.1.1' https://test-site/foo
{}
$ curl https://test-site/foo
access denied
```

I think it needs to check both `XFF` and `RemoteAddr` or it's basically useless.

**Comment by leprechau** [comment]
I've fixed the security hole in the latest commit.

**Comment by leprechau** [comment]
@magiconair I think that's it unless you have other suggestions or requests.

**Comment by magiconair** [comment]
I haven't fully nailed down the test failures. I'm punting by running on travis and codeship and if one of them is green I'm ok with that :-/ I'll have a look at the PR later.

**Comment by leprechau** [comment]
@microadam are you able to test and verify?

**Comment by microadam** [comment]
I will see if I can give it a go! Not a go developer, but will give compiling it a shot

**Comment by leprechau** [comment]
@microadam I could post a binary at my fork of the repo if that makes it easier.  Not sure how comfortable you would be with that though ... and I completely understand.

**Comment by microadam** [comment]
@leprechau yeah that would definitely work! Hitting lots of errors trying to compile (probably just me not knowing how to do it!)

**Comment by leprechau** [comment]
Linux amd64?

**Comment by microadam** [comment]
Have managed to compile it! will try and test it out

**Comment by leprechau** [comment]
Okay, awesome ... let me know if you need anything else.

**Comment by microadam** [comment]
I assume I must be doing something wrong, but with options of `allow=172.14.0.0/12` and accessing it from an address of 192.168.0.x, I don't get denied, which is what I was expecting

I am on Mac OSX running fabio on a Vagrant box, so there are a few networks to contend with

**Comment by leprechau** [comment]
Interesting ... can you turn on access logging just to see what IP fabio sees your connection coming from?  There is probably a NAT between you and the vbox host.  To be really sure you could switch it to  `deny=0.0.0.0/0` that would block everything.

**Comment by microadam** [comment]
No joy with the deny all either. I can only assume I have compiled it wrong. Perhaps if you could provide a binary after all?

Steps to compile were:

```
git clone git@github.com:myENA/fabio.git
git checkout feature/route-acl
go get ./...
make linux
```
I assume something in there is not quite right!

**Comment by microadam** [comment]
Think I have it compiled correctly now. Getting this error:

`[ERROR] failed to process access rules: invalid access item, expected <type>:<data>, got [0.0.0.0/0]`

My `urlprefix` looks like: `urlprefix-my-site.com/ deny=0.0.0.0/0`

**Comment by magiconair** [comment]
@microadam and @leprechau thanks to both of you for working on this!

**Comment by leprechau** [comment]
Also, just realized my own example up there was wrong ... sorry about that.

**Comment by microadam** [comment]
Finally got around to doing some tests, apologies for the delay!

A few things I have spotted, some are possibly edge cases and not worth worrying about (possibly all).

* with `deny=ip:0.0.0.0/0` I can still access it from localhost. Request comes through like:
```
[::1] - - [14/Feb/2018:20:38:10 +0000] "GET / HTTP/1.1" 200 68744
```

* If it put in `allow=ip:192.168.0.252` I correctly get:
```
[ERROR] failed to process access rules: failed to parse CIDR ip:192.168.0.252 with error: invalid CIDR address: 192.168.0.252
```
However it then defaults to allowing all traffic through. As we don't have any concept of "validation" before the server runs, perhaps this should prevent the server from running, or at least, if its an `allow` rule, it should block EVERYTHING, at least you then know something is wrong? Not sure how to handle the case of where its a `deny` rule though, as you wouldn't want to default to ALLOWING everything as you wouldn't know there was an issue. Just trying to think of the case where this is all automated and people aren't paying attention to the logs.

Seems to work in all the other trivial tests I have done! I haven't managed to test it behind another proxy yet, as I don't have that easily setup, but from the looks of the code it should handle that!

Thanks a lot for doing this :)

**Comment by leprechau** [comment]
Thanks for the feedback.  Internally the code is using the golang core `net` package.  Specifically it's using the ParseCIDR and Contains functions.

https://golang.org/pkg/net/#ParseCIDR
https://golang.org/pkg/net/#IPNet.Contains

They support both IPv4 and IPv6 but will not transit boundaries.  So, for the first example with `deny=ip:0.0.0.0/0` that will match ALL IPv4 addresses but will not match IPv6 addresses.  The `[::1]` request source is IPv6 localhost.

The second part is also a side affect of the ParseCIDR function.  `192.168.0.252` is not in proper CIDR notation.  To specify a single IP address in CIDR form it would be `192.168.0.252/32` (a 32 bit mask denotes a single address).

To stopping server operation and/or ignoring the route I have mixed feelings.  The current behavior for invalid options goes ahead and adds the route but just ignores the bad options.  I kept his same behavior for the `allow` and `deny` options.  It is completely possible to either skip the route (not add it) or stop the server if route parsing fails.  However, since routes may be added dynamically I don't think stopping the server for an invalid route option makes any sense.

**Comment by microadam** [comment]
Sure. That all makes sense. 

Just trying to think if there is a way to make it really obvious that it has failed without checking the logs. There isn’t really a clean way of doing it I suppose. I guess if you do mess it up and don’t notice the logs, you will just not see the change you were expecting I.e person still won’t have access or the traffic you are trying to block will still come through. I guess that is the best that can be done!

Definitely a nice addition to this project, certainly covers my use case :)

**Comment by leprechau** [comment]
Awesome, thank you for the feedback and testing.

@magiconair any other thoughts or concerns?  What do you think about skipping the route on failure for allow/deny processing?  Just putting a `continue` in the loop if the process function returns error?

**Comment by magiconair** [comment]
I’ll have a look at the code later after coffee. One thought I had about the addresses is whether to automatically change 1.2.3.4 to 1.2.3.4/32. 

Also, could we add an `all` alias which will just match any address?

Can you do `allow=ip:1.2.3.4/32 deny=all` or is that implicit?

**Comment by leprechau** [comment]
You could check for a string contains and add the slash 32 when there isn’t a match.  I’m not sure an all is needed with the current behavior.  The code currently only allows one of ‘allow’ or ‘deny’ to be specified.  When using the allow option all sources not matched by the allow are denied (default deny).  When using the ‘deny’ option all sources not matched by the deny are allowed (default allow).

**Comment by leprechau** [comment]
I added some code to add a `/32` prefix to value strings that do not contain a `/` ... invalid or incomplete addresses will still error out in the ParseCIDR function.

**Comment by magiconair** [comment]
Should we do this for v6 addresses without netmask as well?

—
Frank Schröder

**Comment by leprechau** [comment]
That's a good question ... would require a bit more pre-parsing but I think it's doable and probably good to be consistent.

**Comment by leprechau** [comment]
Latest commit should properly add a `/32` prefix to plain IPv4 addresses and a `/128` prefix to plain IPv6 addresses.

**Comment by leprechau** [comment]
Last commit added equal test coverage for IPv4 and IPv6 addresses.

File: docs/content/cfg/_index.md [hunk 3]
```diff
@@ -24,14 +24,16 @@ Add a route for a service `svc` for the `src` (e.g. `/path` or `:port`) to a `ds
...
--------------------- | -----------
-`strip=/path`        | Forward `/path/to/file` as `/to/file`
-`proto=tcp`          | Upstream service is TCP, `dst` must be `:port`
-`proto=https`        | Upstream service is HTTPS
-`tlsskipverify=true` | Disable TLS cert validation for HTTPS upstream
-`host=name`          | Set the `Host` header to `name`. If `name == 'dst'` then the `Host` header will be set to the registered upstream host name
-`register=name`      | Register fabio as new service `name`. Useful for registering hostnames for host specific routes.
+Option                                     | Description
+------------------------------------------ | -----------
+`allow=ip:10.0.0.0/8,ip:192.168.0.0/24`    | Restrict access to source addresses within the `10.0.0.0/8` or `192.168.0.0/24` CIDR mask.  All other requests will be denied.
```
Comment (magiconair) [inline docs/content/cfg/_index.md]: Pls add an IPv6 example as well

File: docs/content/feature/access-control.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+---
+title: "Access Control"
+since: "1.5.8"
+---
+
+fabio supports basic ip centric access control per route.  You may
+specify one of `allow` or `deny` options per route to control access.
+Currently only source ip control is available.
```
Comment (magiconair) [inline docs/content/feature/access-control.md]: Pls add an ipv6 example and explain how the source ip is determined. Also add a comment that `/32` and `/128` is added automatically.

File: proxy/http_proxy.go [hunk 5]
```diff
@@ -84,6 +84,11 @@ func (p *HTTPProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if t.AccessDeniedHTTP(r) {
+		http.Error(w, "access denied", http.StatusForbidden)
```
Comment (magiconair) [inline proxy/http_proxy.go]: I'm pretty sure that someone will ask for a different status code here but we can leave it for now.

File: proxy/tcp/sni_proxy.go [hunk 6]
```diff
@@ -78,6 +78,11 @@ func (p *SNIProxy) ServeTCP(in net.Conn) error {
 	}
 	addr := t.URL.Host
 
+	if t.AccessDeniedTCP(in) {
+		log.Print("[INFO] route rules denied access to ", t.URL.String())
```
Comment (magiconair) [inline proxy/tcp/sni_proxy.go]: This should also log who was denied access.

File: proxy/tcp/tcp_proxy.go [hunk 7]
```diff
@@ -48,6 +48,11 @@ func (p *Proxy) ServeTCP(in net.Conn) error {
 	}
 	addr := t.URL.Host
 
+	if t.AccessDeniedTCP(in) {
+		log.Print("[INFO] route rules denied access to ", t.URL.String())
```
Comment (magiconair) [inline proxy/tcp/tcp_proxy.go]: same here

**Comment by leprechau** [comment]
I had to export the `ProcessAccessRules` function that hangs off of `route.Target` in `access_rules.go` in order to process the rules for the integration test.

I also moved the deny logging into the `denyByIP` function in `access_rules.go` to make it easier to alter if needed in the future.

File: docs/content/feature/access-control.md (hunk 4 above)
Comment (leprechau) [reply docs/content/feature/access-control.md]: done!

File: proxy/http_proxy.go (hunk 5 above)
Comment (leprechau) [reply proxy/http_proxy.go]: I can see that being a request but not sure there is much more appropraite than forbidden for a denied request.  Would you want that to be a config option at some point?

Comment (magiconair) [reply proxy/http_proxy.go]: yeah, if someone asks for this.

File: docs/content/feature/access-control.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+Single host addresses (addresses without a prefix) will have a
+`/32` prefix, for IPv4, or a `/128` prefix, for IPv6, added automatically.
+That means `1.2.3.4` is equivalent to `1.2.3.4/32` and `fe80::1234`
+is equivalent to `fe80::1234/128` when specifying
+address blocks for `allow` or `deny` rules.
+
+The source ip used for validation against the defined ruleset is
+taken from information available in the request.
+
+For `HTTP` requests the client `RemoteAddr` is always validated
```
Comment (magiconair) [inline docs/content/feature/access-control.md]: I think if the inbound connection uses the PROXY protocol then this is mapped into the `RemoteAddr` field of the underlying connection. This should be the case for both TCP and HTTP since this is on the connection itself.

https://github.com/fabiolb/fabio/blob/master/vendor/github.com/armon/go-proxyproto/protocol.go#L96-L113

I'll update the docs before merging.

**Comment by magiconair** [comment]
hrmpf, I've merged this but put the additional commit into the branch. This way github won't auto-close the PR ... This is on master now. 

Thanks a lot for @leprechau and @microadam for working on this!
